File: formatter/string_formatter.rb
require_relative 'StatementLexer'
require_relative 'StatementParser'

$lexer_cache = {}
$expression_cache = {}
$debug = false

class StringFormatter

  def initialize(format_string)
    @format_string = format_string
  end

  def pre_compute
    create_expressions
    self
  end

  def evaluate(arguments = nil)
    argument_mapper = ArgumentMap.new(arguments)
    formatted_value = ''
    @expressions.each do |expression|
      formatted_value += expression.evaluate(argument_mapper)
    end
    formatted_value
  end

  class << self

    def find_or_create(format_string, arguments)
      StringFormatter.new(format_string).pre_compute.evaluate(arguments)
    end

  end

  private

  def create_expressions
    if $expression_cache[@format_string]
      @expressions = $expression_cache[@format_string]
    else
      @expressions = []

      index = 0
      str = @format_string
      loop do
        data = str.match(/{{(?<statement>([^}]||}(?!}))*)}}/)
        if data
          index += data.begin(:statement)
          @expressions << Expression.new(data.pre_match, data[:statement], index)
          str = data.post_match
        else
          @expressions << Expression.new(str, nil, index)
          break
        end
      end

      $expression_cache[@format_string] = @expressions
    end
  end

end

class ArgumentMap

  def initialize(arguments)
    @arguments = arguments
  end

  def apply(value)
    return value unless value
    return value unless @arguments
    @arguments.each_with_index do |argument, index|
      value = value.to_s.gsub(/\$#{index + 1}/, argument.to_s)
    end
    value
  end

  def get_value(argument)
    index = argument[1..-1].to_i - 1
    $stderr.puts "parse error #{argument} not found" if $debug and @arguments.size <= index
    @arguments[index]
  end

end

class Expression

  def initialize(pre_match, statement, index)
    @pre_match = pre_match
    @statement = statement
    @index = index
  end

  def evaluate(argument_mapper)
    $argument_mapper = argument_mapper

    parsed_statement = @statement

    unless parsed_statement.nil? || parsed_statement.strip.nil?
      $stderr.puts "parse #{@index}:#{@index + parsed_statement.size}" if $debug

      lexer = create_or_find_lexer(parsed_statement)
      parser = Statement::Parser.new(lexer)
      parsed_statement = parser.program.value
    end

    "#{argument_mapper.apply(@pre_match)}#{argument_mapper.apply(parsed_statement)}"
  end

  private

  def create_or_find_lexer(statement)
    if $lexer_cache[statement]
      lexer = $lexer_cache[statement]
    else
      lexer = Statement::Lexer.new(statement)
      $lexer_cache[statement] = lexer
    end
    lexer.reset
    lexer
  end

end

File: generators/christian/tests/module/mock/src/formatter/StringFormatter.java
package formatter;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.antlr.v4.runtime.ANTLRInputStream;
import org.antlr.v4.runtime.CommonTokenStream;

public class StringFormatter {
	
	public static HashMap<String, CommonTokenStream> tokensMap;
	public static HashMap<String, List<Expression>> expressionsMap;
	
	class Expression {
		
		private String preMatch;
		private String statement;
		private int index;

		public Expression(String preMatch, String statement, int index) {
			this.preMatch = preMatch;
			this.statement = statement;
			this.index = index;
		}
		
		public String evaluate() throws Exception {
			String parsedStatement = statement;
			if (parsedStatement != null && !"".equals(parsedStatement.trim())) {
				StatementParser parser = new StatementParser(createOrFindLexer(parsedStatement));
				parsedStatement = parser.program().value;
			}
			String value = ArgumentMap.apply(StringFormatter.this, preMatch) + ArgumentMap.apply(StringFormatter.this, parsedStatement);
			return value;
		}
		
		private CommonTokenStream createOrFindLexer(String statement) {
			CommonTokenStream tokens;
			if (tokensMap.get(statement) != null) {
				tokens = tokensMap.get(statement);
			} else {
				StatementLexer lexer = new StatementLexer(new ANTLRInputStream(statement));
	            tokens = new CommonTokenStream(lexer);
	            tokens.fill();
	            
	            tokensMap.put(statement, tokens);
			}
			tokens.reset();
            return tokens;
		}
	}

	private String formatString;
	private List<Expression> expressions;
	
	public StringFormatter(String formatString) {
		this.formatString = formatString;

    if (tokensMap == null)
      tokensMap = new HashMap<String, CommonTokenStream>();
    if (expressionsMap == null)
      expressionsMap = new HashMap<String, List<Expression>>();
	}
	
	public StringFormatter preCompute() {
		createExpressions();
		return this;
	}
	
	public String evaluate(String[] arguments) throws Exception {
		String formattedValue = "";
		try {
			ArgumentMap.setArguments(this, arguments);		
			for (Expression expression : expressions) {
				formattedValue += expression.evaluate();
			}
		} finally {
			ArgumentMap.removeArguments(this);
		}
		return formattedValue;
	}
	
	private void createExpressions() {
		if (expressionsMap.get(formatString) != null) {
			expressions = expressionsMap.get(formatString);
		} else {
			expressions = new ArrayList<Expression>();
			
			int index = 0;
			int lastEnd = 0;
			String str = formatString;
			Pattern pattern = Pattern.compile("\\{\\{(([^\\}]||\\}(?!\\}))*)\\}\\}");
			Matcher matcher = pattern.matcher(str);
			while (true) {
				if (lastEnd < str.length() && matcher.find(lastEnd)) {
					index = matcher.start();
					String preMatch = "";
					if (matcher.start() > 0) {
						preMatch = str.substring(lastEnd, matcher.start());
					}
					String statement = str.substring(matcher.start() + 2, matcher.end() - 2);
					expressions.add(new Expression(preMatch, statement, index));
					lastEnd = matcher.end();
				} else {
					String preMatch = "";
					if (lastEnd < str.length()) {
						preMatch = str.substring(lastEnd);
					}
					expressions.add(new Expression(preMatch, null, index));
					break;
				}
			}
			
			expressionsMap.put(formatString, expressions);
		}
	}
	
	public static void init() {
		tokensMap = new HashMap<String, CommonTokenStream>();
		expressionsMap = new HashMap<String, List<Expression>>();
	}
}
